src/wasmtime/component.rs

```diff
@@ -107,7 +107,60 @@ impl WasmValue for component::Val {
     type Type = component::Type;
 
     fn ty(&self) -> Self::Type {
-        self.ty()
+        match self {
+            Self::Bool(_) => component::Type::Bool,
+            Self::S8(_) => component::Type::S8,
+            Self::U8(_) => component::Type::U8,
+            Self::S16(_) => component::Type::S16,
+            Self::U16(_) => component::Type::U16,
+            Self::S32(_) => component::Type::S32,
+            Self::U32(_) => component::Type::U32,
+            Self::S64(_) => component::Type::S64,
+            Self::U64(_) => component::Type::U64,
+            Self::Float32(_) => component::Type::Float32,
+            Self::Float64(_) => component::Type::Float64,
+            Self::Char(_) => component::Type::Char,
+            Self::String(_) => component::Type::String,
+            _ => unimplemented!(),
+            /*
+            Self::List(_) => component::Type::List ...,
+            Self::Record(_) => component::Type::Record ...,
+            Self::Tuple(_) => component::Type::Tuple ...,
+            Self::Variant(_, _) => component::Type::Variant ...,
+            Self::Enum(_) => component::Type::Enum ...,
+            Self::Option(_) => component::Type::Option ...,
+            Self::Result(_) => component::Type::Result ...,
+            Self::Flags(_) => component::Type::Flags ...,
+            Self::Resource(_) => WasmTypeKind::Unsupported,
+             */
```

Comment:
So the tests don't run into the unimplemented cases. The `write` call uses the new `kind` method directly.

Maybe the idea of going to a Type can be changed and the logic can go directly to a Kind. That would be a more dramatic change.

Do you want to hold off reviewing the other changes until this is cleaned up? I drew inspiration from the original having used unimplemented for the more complicated types too. And do you want to do this now or is there time for me to try and figure out some more things?

Something that could help me if you think there is benefit in me trying to resolve this one, are there other unit tests you would like to seen work that aren't there yet?

Reply:
Sorry I haven't had a chance to look at exactly what is going on here; I'll try to find time tomorrow.

Reply:
When you do look, notice that the use of val.ty().kind() has been replaced with val.kind() in two places - I think that is the interesting part. So the ty() trait method wasn't in play in the two critical areas.

Reply:
OK, I have a better grasp on the problem now; here's what I'm thinking:

First, replace `WasmValue::ty` with `::kind`. As you note, some uses of `ty` really only need the kind.

The remaining uses of `ty` are for type checking, which this PR works around by relaxing those checks to only validate `kind`. I'd like to get that type checking back by add a new method to `WasmValue`: `fn is(&self, ty: &Self::Type) -> bool`. I believe it should be relatively straightforward to implement this for `component::Val` though it may take a fair bit of code to handle all the composite types.

Does that sound reasonable? If you are willing to attempt the above in this PR that would be great, otherwise I'm fine merging this as-is and will probably try to make the other changes before the next release.

Reply:
> The remaining uses of ty are for type checking, which this PR works around by relaxing those checks to only validate kind. 

While I don't know *exactly* what you mean that some checks are relaxed. If you mean some calls to `ty` for a component::Val would result in a panic with the current PR commits, I agree. I wasn't sure if they were important to handle because I didn't see a unit test fail.

Would it take new unit tests to show them fail or some other API for the wasm-wave library itself that isn't fleshed out yet?

I would rather take on this next step if I could see it actually passing a test as a result.

So I can start trying to figure out the next step from your guidance, but if the current PR doesn't make things worse, I would prefer this go in, and other improvements go into a new PR. Makes understanding and possible git bisections later easier.

Reply:
> I don't know exactly what you mean that some checks are relaxed. 

https://github.com/lann/wasm-wave/pull/34/files#r1601813873

Reply:
The test coverage you're looking for is missing because the test cases can't be constructed just from the wasmtime API; you have to load an actual wasm file and reflect on its types to get the relevant test fixtures.

Reply:
...though immediately after typing that I remembered that I implemented exactly that test fixture in `tests/wasmtime.rs`, so I probably can just expand coverage there.

Reply:
(I'd still be fine merging this PR as-is if you prefer)

Reply:
Thanks for getting back to me on this. I've been following the idea that a bad value for a Flags type was being caught with wasmtime 18 and now it isn't, at least the way I had changed things.

And sure enough, a test based on the tests/wasmtime/test_round_trips_unchanged does fail with I give it a garbage flag name. I'm in the process of figuring out how the wasmtime library knew that was not value.

Reply:
Okay. I found the wasmtime 20 library has a 'names()' method on the Flags type. So it would be straightforward to perform the check manually in that make_flags method.

Two questions. Should the error message have both the invalid string and the possible legal ones? I don't think the original gave that much information. And I wonder that the wasmtime library doesn't provide the validation check any longer. Does its author want each embedder to code their own checks?

Reply:
Most wasmtime embedders use `wit-bindgen` which enforces types through code generation; they don't use this Val interface at all. I believe the API change allowed the implementation to be simplified quite a bit at a (small) expense to weird crates like this one :slightly_smiling_face:

Reply:
Sorry, realized I didn't respond to one of your questions:

> Should the error message have both the invalid string and the possible legal ones? I don't think the original gave that much information.

I'm fine with either in this PR; nicer errors can always be added later.